# Data Collection

## Project: AI-Assisted Pull Request Outcome Prediction

In this step, we collect and organize the real-world data required for our machine learning project.

Our project uses AI-assisted Pull Request data collected from GitHub through the AIDev dataset.

The required AIDev data is collected in Parquet format and includes Pull Request, repository, and user information.

The Pull Request dataset contains a large number of records, so Parquet format allows us to store and work with the data efficiently.

In [25]:
# Import libraries required for working with our Parquet dataset

import pyarrow.parquet as pq
import pandas as pd

In [26]:
# Locate and open the original Pull Request dataset

file_path = "../data/raw/all_pull_request.parquet"

file = pq.ParquetFile(file_path)

In [27]:
# Check the total rows, row groups, and columns in the dataset

print("Number of rows:", file.metadata.num_rows)
print("Number of row groups:", file.num_row_groups)
print("Number of columns:", len(file.schema.names))

Number of rows: 2743854
Number of row groups: 3
Number of columns: 14


In [28]:
# Display all columns available in the Pull Request dataset

print("Columns:")
for column in file.schema.names:
    print(column)

Columns:
id
number
title
body
agent
user_id
user
state
created_at
closed_at
merged_at
repo_id
repo_url
html_url


In [29]:
# Read a small sample instead of loading the entire 1.5 GB dataset

sample = file.read_row_group(0).to_pandas()
sample = sample.head(10000)

print("Sample shape:", sample.shape)

Sample shape: (10000, 14)


In [30]:
# View the first few Pull Request records

sample.head()

,id,number,title,body,agent,user_id,user,state,created_at,closed_at,merged_at,repo_id,repo_url,html_url
0,3368366527,51,Pull request for issue #33,Fixes #33,Google_Jules,161369871,google-labs-jules[bot],closed,2025-08-29T22:33:16Z,2025-08-29T22:34:18Z,2025-08-29T22:34:18Z,1.046176e+09,https://api.github.com/repos/devhitoshi/call-g...,https://github.com/devhitoshi/call-generator/p...
1,3368370520,21,feat: Add comprehensive unit tests for backend...,This change adds a comprehensive suite of unit...,Google_Jules,161369871,google-labs-jules[bot],closed,2025-08-29T22:34:52Z,2025-08-30T18:40:38Z,2025-08-30T18:40:38Z,1.035723e+09,https://api.github.com/repos/s4lly/splitzy,https://github.com/s4lly/splitzy/pull/21
2,3368379454,1,Modularize Portfolio CLI Component,This change modularizes the single-file `Portf...,Google_Jules,161369871,google-labs-jules[bot],closed,2025-08-29T22:39:10Z,2025-08-31T00:32:34Z,NaN,1.046866e+09,https://api.github.com/repos/papadonut9/cli-po...,https://github.com/papadonut9/cli-portfolio-we...
3,3368383201,84,docs: Create UI component guide,"This change adds a new documentation file, `UI...",Google_Jules,161369871,google-labs-jules[bot],closed,2025-08-29T22:40:55Z,2025-08-29T22:54:33Z,2025-08-29T22:54:33Z,1.015524e+09,https://api.github.com/repos/torus-automations...,https://github.com/torus-automations/groupweav...
4,3368385471,1,Update README with setup and installation inst...,This change updates the README.md file to be m...,Google_Jules,161369871,google-labs-jules[bot],closed,2025-08-29T22:41:45Z,2025-08-29T22:42:51Z,2025-08-29T22:42:51Z,1.044730e+09,https://api.github.com/repos/OmniFlexFitness/C...,https://github.com/OmniFlexFitness/Colab-Image...


In [31]:
# Check the sample's columns, data types, and non-null values

sample.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          10000 non-null  int64  
 1   number      10000 non-null  int64  
 2   title       10000 non-null  str    
 3   body        9996 non-null   str    
 4   agent       10000 non-null  str    
 5   user_id     10000 non-null  int64  
 6   user        10000 non-null  str    
 7   state       10000 non-null  str    
 8   created_at  10000 non-null  str    
 9   closed_at   8803 non-null   str    
 10  merged_at   7778 non-null   str    
 11  repo_id     9998 non-null   float64
 12  repo_url    10000 non-null  str    
 13  html_url    10000 non-null  str    
dtypes: float64(1), int64(3), str(10)
memory usage: 7.7 MB


---

In [6]:
# Locate and open the original AIDev repository metadata dataset.

repository_path = "../data/raw/all_repository.parquet"

repository_file = pq.ParquetFile(repository_path)

In [7]:
# Check the total rows, row groups, and columns in the repository dataset.

print("Number of rows:", repository_file.metadata.num_rows)
print("Number of row groups:", repository_file.num_row_groups)
print("Number of columns:", len(repository_file.schema.names))

Number of rows: 326798
Number of row groups: 1
Number of columns: 8


In [8]:
# Display all columns available in the repository dataset.

print("Repository columns:")
for column in repository_file.schema.names:
    print(column)

Repository columns:
id
url
license
full_name
is_forked
language
forks
stars


In [9]:
# Read a small repository sample for inspection.

repository_sample = repository_file.read_row_group(0).to_pandas()
repository_sample = repository_sample.head(10000)

print("Repository sample shape:", repository_sample.shape)

Repository sample shape: (10000, 8)


In [10]:
# View the first few repository records.

repository_sample.head()

,id,url,license,full_name,is_forked,language,forks,stars
0,976907397,https://api.github.com/repos/tonionagauzzi/Wea...,NaN,tonionagauzzi/WeatherApp,False,Kotlin,0.0,0.0
1,1012708699,https://api.github.com/repos/Gitlippe/DGM,Apache-2.0,Gitlippe/DGM,True,Python,0.0,0.0
2,1024669864,https://api.github.com/repos/oyi77/telegram-ai...,NaN,oyi77/telegram-ai-trade,False,Python,0.0,0.0
3,1053420190,https://api.github.com/repos/felipebehling/Har...,MIT,felipebehling/HardStress,False,C,0.0,1.0
4,1077676076,https://api.github.com/repos/notleks2411/gnome...,GPL-3.0,notleks2411/gnome-abrt,True,Python,0.0,0.0


In [11]:
# Check the repository sample's columns, data types, and non-null values.

repository_sample.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         10000 non-null  int64  
 1   url        10000 non-null  str    
 2   license    3372 non-null   str    
 3   full_name  10000 non-null  str    
 4   is_forked  10000 non-null  bool   
 5   language   8459 non-null   str    
 6   forks      10000 non-null  float64
 7   stars      10000 non-null  float64
dtypes: bool(1), float64(2), int64(1), str(4)
memory usage: 1.4 MB


---

In [12]:
# Locate and open the original AIDev user metadata dataset.

user_path = "../data/raw/all_user.parquet"

user_file = pq.ParquetFile(user_path)

In [13]:
# Check the total rows, row groups, and columns in the user dataset.

print("Number of rows:", user_file.metadata.num_rows)
print("Number of row groups:", user_file.num_row_groups)
print("Number of columns:", len(user_file.schema.names))

Number of rows: 161255
Number of row groups: 1
Number of columns: 5


In [14]:
# Display all columns available in the user dataset.

print("User columns:")
for column in user_file.schema.names:
    print(column)

User columns:
id
login
followers
following
created_at


In [15]:
# Read a small user sample for inspection.

user_sample = user_file.read_row_group(0).to_pandas()
user_sample = user_sample.head(10000)

print("User sample shape:", user_sample.shape)

User sample shape: (10000, 5)


In [16]:
# View the first few user records.

user_sample.head()

,id,login,followers,following,created_at
0,51655670.0,jbgolds,0.0,0.0,2019-06-10T14:06:23Z
1,100471171.0,WalkieCat,0.0,1.0,2022-02-26T11:21:00Z
2,210220192.0,milanjethva-09,0.0,0.0,2025-05-04T06:57:54Z
3,89417141.0,nfantinodyer,3.0,4.0,2021-08-23T18:58:54Z
4,8058221.0,jankon,0.0,0.0,2014-07-03T12:22:06Z


In [17]:
# Check the user sample's columns, data types, and non-null values.

user_sample.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          9994 non-null   float64
 1   login       10000 non-null  str    
 2   followers   9994 non-null   float64
 3   following   9994 non-null   float64
 4   created_at  9994 non-null   str    
dtypes: float64(3), str(2)
memory usage: 687.7 KB


---

# Saving Sample Data

In [33]:
# Save inspection samples from the Pull Request, Repository, and User datasets as Excel files.

sample.to_excel(
    "../data/sample/pr_sample.xlsx",
    index=False
)

repository_sample.to_excel(
    "../data/sample/repository_sample.xlsx",
    index=False
)

user_sample.to_excel(
    "../data/sample/user_sample.xlsx",
    index=False
)

print("PR sample saved successfully.")
print("Repository sample saved successfully.")
print("User sample saved successfully.")

C:\Users\srich\AppData\Local\Temp\ipykernel_19908\1864610418.py:3: UserWarning: Cell contents too long (34897), truncated to 32767 characters
  sample.to_excel(
C:\Users\srich\AppData\Local\Temp\ipykernel_19908\1864610418.py:3: UserWarning: Cell contents too long (36099), truncated to 32767 characters
  sample.to_excel(


PR sample saved successfully.
Repository sample saved successfully.
User sample saved successfully.


---

## Data Collection Conclusion

The AIDev datasets were successfully collected and stored in Parquet format.

- The Pull Request dataset contains **2,743,854 Pull Requests** and **14 columns**.
- The Repository dataset contains **326,798 repositories** and **8 columns**.
- The User dataset contains **161,255 user records** and **5 columns**.
- Small samples from all three datasets were saved as Excel files for inspection.
- The original datasets were kept unchanged.
- The collected datasets will be used in later stages for data validation, integration, and machine learning preparation.

---